File: src/2017/day02/typescript/solution.ts
function formatInput(input: string): number[][] {
  return input.split('\n').map((row) => row.split(/\s+/g).map(Number));
}

export function part1(input: string): number | string | undefined {
  const rows = formatInput(input);
  const differences = rows.map((row) => Math.max(...row) - Math.min(...row));
  return differences.reduce((sum, num) => sum + num);
}

function getEvenDivision(row: number[]): number {
  for (let l = 0; l < row.length; l++) {
    for (let r = l + 1; r < row.length; r++) {
      if (row[l] % row[r] === 0) {
        return row[l] / row[r];
      }
      if (row[r] % row[l] === 0) {
        return row[r] / row[l];
      }
    }
  }
  return 0;
}

export function part2(input: string): number | string | undefined {
  const divisors = formatInput(input).map((row) => getEvenDivision(row));
  return divisors.reduce((sum, divisor) => sum + divisor);
}

File: src/2017/day02/python/solution.py
def format_input(input: str) -> list[list[int]]:
    return [[int(num) for num in line.split()] for line in input.strip().splitlines()]


def part1(input: str) -> int:
    rows = format_input(input)
    differences = [max(row) - min(row) for row in rows]
    return sum(differences)


def get_even_division(row: list[int]) -> int:
    for left_index in range(len(row)):
        for right_index in range(left_index + 1, len(row)):
            if row[left_index] % row[right_index] == 0:
                return int(row[left_index] / row[right_index])
            if row[right_index] % row[left_index] == 0:
                return int(row[right_index] / row[left_index])
    return 0


def part2(input: str) -> int:
    divisors = [get_even_division(row) for row in format_input(input)]
    return sum(divisors)
